Cpp14 support

oliverlee opened the pull request:
I've updated the P0009 headers to compile without C++17 extensions.

For the last commit, I've dumped a few common utilities in the extents header. Please feel free to edit as desired.

oliverlee committed:
replace fold expressions in `extents.hpp` with macro

Use the `_MDSPAN_FOLD_AND` macro to avoid using fold expressions
when building for C++14.

oliverlee committed:
use empty message with static assert

Add an empty message to static assert in `layout_padded_fwd.hpp` to
avoid use of C++17 extensions when building for C++14.

oliverlee committed:
replace constexpr if in p0009 bits

Replace use of constexpr if in headers used to implement p0009 in
order to allow compilation for C++14 without C++17 extensions.

mhoemmen commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -16,11 +16,14 @@
 #endif
 #include <array>
+#include <type_traits>
+#include <stdexcept>
```
Nothing in this header threw an `std::exception` subclass before.  Furthermore, the only reason this code now throws an exception is to trigger a call to `terminate` in a `noexcept` function.

Would you consider just calling `std::abort` as before?  Alternately, would you consider adding a configurable precondition violation handler (e.g., a macro that takes a message and just calls `std::abort` by default)?

oliverlee replied:
> Would you consider just calling std::abort as before? Alternately, would you consider adding a configurable precondition violation handler (e.g., a macro that takes a message and just calls std::abort by default)?

Sure that sounds like a better approach. I'm also wondering if `extents.hpp` is the right location for the `terminate_if_invalid_strides` function. Maybe `layout_stride.hpp` is a better place?

mhoemmen commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -367,6 +370,82 @@ struct maybe_static_array {
+template <std::size_t N, class Layout, class Extents, class Mapping>
+constexpr void terminate_if_invalid_strides(with_rank<N>, Layout, const Extents& ext, const Mapping& other)
+{
+  static_assert(std::is_same<Layout, layout_left>::value or
+                std::is_same<Layout, layout_right>::value
+                , "");
```
Would you consider adding a nonempty message here?

mhoemmen commented on include/experimental/__p0009_bits/fwd.hpp:
```diff
@@ -0,0 +1,24 @@
+//
+// this is used to enable alternatives to constexpr if when building for C++14
+//
+template <std::size_t N>
+struct with_rank {};
```
Would you consider making this an alias of `std::integral_constant`, so as to avoid more instantiations of types?

mhoemmen commented:
Thanks for the contribution!  There are some nice refactorings in there as well.  @crtrott et al. will likely want to comment.

1. Do mdspan's tests and examples all build with C++14, or do these changes only just make your application build?  We're generally only interested in the former.

2. Would you consider adding C++14 testing to https://github.com/kokkos/mdspan/blob/stable/.github/workflows/cmake.yml as a part of this PR?  Current PR automatic testing will not exercise C++14.

oliverlee commented:
>     1. Do mdspan's tests and examples all build with C++14, or do these changes only just make your application build?  We're generally only interested in the former.

I've only tested with the tests defined in CMake along with C++14 so far. All tests pass before and after this PR - however after this PR now builds without warnings.

>     2. Would you consider adding C++14 testing to https://github.com/kokkos/mdspan/blob/stable/.github/workflows/cmake.yml as a part of this PR?  Current PR automatic testing will not exercise C++14.

Sure.

oliverlee committed:
ci: test with C++14 and treat compile warnings as errors

oliverlee committed:
define precondition macro and violation handler

mhoemmen commented:
Thanks for making the changes!  Building with C++14 results in the following CMake error message.

```
CMake Error at CMakeLists.txt:200 (MESSAGE):
  Benchmarks are not available in C++14 or C++23 mode.  Turn
  MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.
```

I think you could just change line 200 of CMakeLists.txt ( https://github.com/kokkos/mdspan/blob/0e6a69dfe045acbb623003588a4aff844ea4b276/CMakeLists.txt#L200 ) from what it is currently,

```
MESSAGE(FATAL_ERROR "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
```

to something like

```
set(MDSPAN_ENABLE_BENCHMARKS OFF)
MESSAGE(STATUS "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
```

oliverlee commented:
> Thanks for making the changes! Building with C++14 results in the following CMake error message.
> 
> ```
> CMake Error at CMakeLists.txt:200 (MESSAGE):
>   Benchmarks are not available in C++14 or C++23 mode.  Turn
>   MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.
> ```
> 
> I think you could just change line 200 of CMakeLists.txt (
> 
> https://github.com/kokkos/mdspan/blob/0e6a69dfe045acbb623003588a4aff844ea4b276/CMakeLists.txt#L200
> ) from what it is currently,
> 
> ```
> MESSAGE(FATAL_ERROR "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
> ```
> 
> to something like
> 
> ```
> set(MDSPAN_ENABLE_BENCHMARKS OFF)
> MESSAGE(STATUS "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
> ```

I disabled benchmarking for C++14 using the same approach for C++23 (i.e. via matrix settings in `cmake.yml`) since I thought it would be simpler to maintain. Let me know if you still prefer a CMake option based approach.

crtrott commented on include/experimental/__p0009_bits/extents.hpp:
```diff
@@ -537,14 +539,9 @@ template <class IndexType, size_t... Extents> class extents {
-    }
-    return true;
+    return
+      rank() == extents<OtherIndexType, OtherExtents...>::rank() and
```
```suggestion
      rank() == extents<OtherIndexType, OtherExtents...>::rank() &&
```

crtrott commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+} // namespace detail
+} // namespace MDSPAN_IMPL_STANDARD_NAMESPACE
+
+#ifndef MDSPAN_PRECONDITION_VIOLATION_HANDLER
+#define MDSPAN_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
```
```suggestion
#ifndef MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER
#define MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
```

crtrott commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+  MDSPAN_IMPL_STANDARD_NAMESPACE::detail::default_precondition_violation_handler(cond, file, line)
+#endif
+
+#ifndef MDSPAN_SKIP_PRECONDITION_VIOLATION_HANDLER
+  #ifndef NDEBUG
+    #define MDSPAN_SKIP_PRECONDITION_VIOLATION_HANDLER 0
+  #else
+    #define MDSPAN_SKIP_PRECONDITION_VIOLATION_HANDLER 1
+  #endif
+#endif
```
```suggestion
#ifndef MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER
  #ifndef NDEBUG
    #define MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER 0
  #else
    #define MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER 1
  #endif
#endif
```

crtrott commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+namespace MDSPAN_IMPL_STANDARD_NAMESPACE {
+namespace detail {
+
+template <bool skip = MDSPAN_SKIP_PRECONDITION_VIOLATION_HANDLER>
```
```suggestion
template <bool skip = MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER>
```

crtrott commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+} // namespace detail
+} // namespace MDSPAN_IMPL_STANDARD_NAMESPACE
+
+#define MDSPAN_PRECONDITION(...) \
```
```suggestion
#define MDSPAN_IMPL_PRECONDITION(...) \
```

crtrott commented on include/experimental/__p0009_bits/layout_stride.hpp:
```diff
@@ -560,4 +574,34 @@ struct layout_stride {
+  for (std::size_t r = 0; r < N; r++) {
+    const std::size_t s = is_left ? r : N - 1 - r;
+
+    MDSPAN_PRECONDITION(common_integral_compare(stride, other.stride(s))
+                        and "invalid strides for layout_{left,right}");
```
```suggestion
    MDSPAN_IMPL_PRECONDITION(common_integral_compare(stride, other.stride(s))
                             and "invalid strides for layout_{left,right}");
```

crtrott commented on tests/test_alternate_precondition_violation_handler.cpp:
```diff
@@ -0,0 +1,29 @@
+#include <stdexcept>
+
+#define MDSPAN_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
+  do { \
+    throw std::logic_error{"precondition failure"}; \
+  } while (0);
```
```suggestion
#define MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
  do { \
    throw std::logic_error{"precondition failure"}; \
  } while (0);
```

crtrott commented on tests/test_alternate_precondition_violation_handler.cpp:
```diff
@@ -0,0 +1,29 @@
+
+TEST(mdspan_macros, alternate_precondition_violation_handler)
+{
+  ASSERT_THROW(MDSPAN_PRECONDITION(false), std::logic_error);
```
```suggestion
  ASSERT_THROW(MDSPAN_IMPL_PRECONDITION(false), std::logic_error);
```

crtrott commented on tests/test_alternate_precondition_violation_handler.cpp:
```diff
@@ -0,0 +1,29 @@
+  {
+    constexpr auto operator()() const
+    {
+      MDSPAN_PRECONDITION(1 + 1 == 2);
```
```suggestion
      MDSPAN_IMPL_PRECONDITION(1 + 1 == 2);
```

crtrott commented on tests/test_macros.cpp:
```diff
@@ -0,0 +1,24 @@
+{
+  constexpr auto msg = "hello, world!";
+
+  ASSERT_DEATH(MDSPAN_PRECONDITION(false and "hello, world!"), msg);
```
```suggestion
  ASSERT_DEATH(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"), msg);
```

crtrott commented on tests/test_macros.cpp:
```diff
@@ -0,0 +1,24 @@
+  {
+    constexpr auto operator()() const
+    {
+      MDSPAN_PRECONDITION(1 + 1 == 2);
```
```suggestion
      MDSPAN_IMPL_PRECONDITION(1 + 1 == 2);
```

crtrott reviewed (changes requested):
This looks pretty good. Finally found sufficient time to review it. The new macros are not a bad idea, but I think they should all be IMPL macros.

crtrott commented:
I will rebase this and merge my suggestions.

crtrott committed:
Make new precondition macros IMPL only

crtrott commented:
@oliverlee FYI: I rebased and force pushed. I'd like to get this in before I work on more of the features for C++26.

crtrott committed:
Make test for dims C++14 compatible

oliverlee commented:
Not sure why the death tests are failing now - they were fine a few months back.

I'll probably have time to look into it tomorrow.

crtrott commented:
They passed now. We might have changed Debug to Release build at some point and so the rebase made them fail.

oliverlee commented:
It should be possible to enable precondition checks in release mode with a preprocessor macro - I imagine this is enabled only for tests and disabled for benchmarks.

It shouldn't be too much trouble to do in a separate PR if you want to merge this PR and use it as a base for c++26 stuff.

oliverlee commented:
@crtrott It looks like the macro in `test/CMakeLists.txt` was missed in the rename commit.

I've created two commits - please squash or do whatever is normal for this repo.

crtrott commented:
Thanks! I had overlooked that. That said I am not sure it's the best way to go about this. Now the tests never take the code path where the preconditions are skipped. Maybe instead we should have a mix of debug and non-debug test configs.

crtrott commented:
Also I am waiting for a second review to finish (should happen latest by Monday I pinged appropriate people in my team).

oliverlee commented:
> That said I am not sure it's the best way to go about this. 

Same - whatever makes the most sense to you in terms of maintenance.

I've seen precondition tests defined in a separate test file. That could remove the need to conditionally define test cases with `ASSERT_DEATH` and simplify running tests with debug and non-debug configs. Would that make sense here?

crtrott commented:
It could make sense to separate them out. We had some platforms where the death tests never fully worked ...

oliverlee commented:
> It could make sense to separate them out. We had some platforms where the death tests never fully worked ...

There are only a few `DEATH_TESTS` now.

It makes sense to do it earlier rather than later if this is approach you want to pursue.

crtrott commented:
Yeah let's do it. Do you have time to tackle that? If not I might get around to it some point this weekend.

oliverlee commented:
> Yeah let's do it. Do you have time to tackle that? If not I might get around to it some point this weekend.

Maybe this afternoon? If not, also this weekend.

crtrott commented:
Sounds good. Also I send you an invite to our slack channel in case you are interested.

crtrott commented:
Note: I merged one other PR, but this one rebases cleanly on top of it.

oliverlee committed:
replace `std::is_signed_v` for C++14 compatibility

oliverlee committed:
isolate precondition tests

Move test cases that test preconditions with `ASSERT_DEATH` to separate
test files. This allows other (non-precondition) tests to be compiled
with or without preconditions. CMake test targets will enable
preconditions during compilation if passed the option
`ENABLE_PRECONDITIONS`.

This only requires moving a single test case from `test_layout_ctors` -
`test_alternate_precondition_violation_handler`
and `test_macros` are test files have been written primary to test
preconditions.

oliverlee commented:
Rebased + squashed some commits + separated out the death tests. Is this what you had in mind?

crtrott commented:
Yeah this is great. Thanks so much!

dalg24 commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+namespace detail {
+
+template <bool check = MDSPAN_IMPL_CHECK_PRECONDITION>
+constexpr void precondition(const char* cond, const char* file, unsigned line)
```
I think this needs to be decorated with `MDSPAN_FUNCTION`

dalg24 replied:
Did you consider tag dispatching instead ,having two non templated overloads
```C++
MDSPAN_FUNCTION constexpr void precondition(std::false_type /*checked*/, const char* /*cond*/, const char* /*file*/, unsigned /*line*/) { /* do nothing */ }
MDSPAN_FUNCTION constexpr void precondition(std::true_type /*checked*/, const char* cond, const char* file, unsigned line) { MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line); }

oliverlee replied:
I might have a while back? Not sure.

But I tested it out just now, and this approach produces warnings with Clang and GCC:

Clang-18:
```
.../mdspan/include/mdspan/../experimental/__p0009_bits/macros.hpp:142:16: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
  142 | constexpr void precondition(std::true_type, const char* cond, const char* file, unsigned line)
```

GCC-11:
```
.../mdspan/include/mdspan/../experimental/__p0009_bits/macros.hpp:120:81: error: call to non-‘constexpr’ function ‘void Kokkos::detail::default_precondition_violation_handler(const char*, const char*, unsigned int)’
```

dalg24 replied:
I expect the issue is the `constexpr` specifier

dalg24 replied:
I suppose you needed it to be call it in other functions that themselves are callable in a constant expression...

dalg24 commented on include/experimental/__p0009_bits/utility.hpp:
```diff
@@ -0,0 +1,62 @@
+  return match;
+}
+
+constexpr struct
+{
+  template <class T, class I>
+  constexpr auto operator()(const T& x, I i) const
+  {
+    return x.extent(i);
+  }
+} extent;
+
+constexpr struct
+{
+  template <class T, class I>
+  constexpr auto operator()(const T& x, I i) const
+  {
+    return x.stride(i);
+  }
+} stride;
```
Where are these used?

crtrott replied:
line 526 in layout_stride.hpp and 546 or so in extents.hpp respectively.

dalg24 commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+namespace MDSPAN_IMPL_STANDARD_NAMESPACE {
+namespace detail {
+
+inline void default_precondition_violation_handler(const char* cond, const char* file, unsigned line)
```
Fine by me but just noting that the default handler would not work with CUDA/HIP

oliverlee replied:
Please feel free to make any changes if CUDA/HIP compatibility is necessary.

mhoemmen replied:
`printf` works just fine with CUDA; it's HIP that's broken lol ; - )

dalg24 replied:
it is a `fprintf` and it is qualified so there is no way it works on CUDA.  As far as I know HIP behaves just the same as CUDA here.

dalg24 commented on include/experimental/__p2642_bits/layout_padded_fwd.hpp:
```diff
@@ -82,36 +83,49 @@ struct is_layout_right_padded_mapping<_Mapping,
+    layout_padded_constants<typename _OtherMapping::layout_type,
+                            _ExtentsType>::padded_stride_idx;
+  constexpr auto extent_to_pad_idx = layout_padded_constants<typename _OtherMapping::layout_type, _ExtentsType>::extent_to_pad_idx;
+  assert(other_mapping.stride(padded_stride_idx) == other_mapping.extents().extent(extent_to_pad_idx));
```
Should we not be using `MDSPAN_IMPL_PRECONDITION` here?

oliverlee replied:
Sure let's do that.

dalg24 commented on tests/CMakeLists.txt:
```diff
@@ -6,14 +6,24 @@ if(MDSPAN_ENABLE_CUDA)
   target_link_libraries(${name} mdspan gtest_main)
   add_test(${name} ${name})
-endmacro()
+  set_property(TARGET ${name} PROPERTY COMPILE_WARNING_AS_ERROR ON)
```
I am not necessarily opposed to this but IMO it does not belong to this PR

oliverlee replied:
This change is related to C++14 support to some degree - otherwise C++17 language features are still usable with some compilers when building for C++14. But I also see that the impact of this change has wider consequences.

What do you think about breaking out this change into a separate PR and rebasing this PR on top of that? I can do that later this evening - or feel free to change this PR as you see fit.

oliverlee replied:
https://github.com/kokkos/mdspan/pull/328

dalg24 replied:
Christian did you mean to keep this?

dalg24 commented on tests/test_alternate_precondition_violation_handler.cpp:
```diff
@@ -0,0 +1,29 @@
+#include <stdexcept>
+
+#define MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
+  do { \
+    throw std::logic_error{"precondition failure"}; \
+  } while (0);
```
I am confused.  What is this suppose to test?
~~For one it is throwing regardless of whether the condition evaluates to false or not.~~ **EDIT: I see now that the condition is meant for debugging purposes, not for evaluation.  Sorry about the noise**
Also the `MDSPAN_IMPL_` denotes that that the macro is "private" or "internal" which indicates it is not meant as a customization point for users.

oliverlee replied:
No worries about the noise. Should there be some documentation here for future readers?

crtrott replied:
The original idea was to have the macros public. But I wanted to consider that a bit more in-depth and thus said to make it private for now. We should still test the mechanism.

dalg24 replied:
Sure

dalg24 commented on tests/test_macros.cpp:
Do we care to test that precondition checking can indeed be disabled?

oliverlee replied:
Yes that's probably a good idea.

I'm not sure what's the best way - possibly a new test file+target that defines a precondition handler that unconditionally throws and then setting `MDSPAN_IMPL_CHECK_PRECONDITION 0`?

mhoemmen commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +103,52 @@
+namespace detail {
+
+inline void default_precondition_violation_handler(const char* cond, const char* file, unsigned line)
+{
+  (void)std::fprintf(::stderr, "%s:%u: precondition failure: `%s`\n", file, line, cond);
+  std::abort();
+}
```
I don't know how HIP or SYCL work, but the following should be fine for CUDA.  Note that in device code, `printf` may have a performance cost even if it is never called at run time.  (For example, the compiler may need to reserve registers for it.)

```suggestion
{
#if defined(__CUDA_ARCH__)
#  if ! defined(NDEBUG)
  printf("%s:%u: precondition failure: `%s`\n", file, line, cond);
  assert(false);
#  endif
#else
  fprintf(stderr, "%s:%u: precondition failure: `%s`\n", file, line, cond);
  assert(false);
#endif 
}
```

masterleinad replied:
`HIP` works the same as `Cuda`. For SYCL (oneAPI), you would need to use `sycl::ext::oneapi::experimental::printf`.

oliverlee committed:
use MDSPAN_IMPL_PRECONDITION in layout_padded_fwd

oliverlee committed:
test that precondition checks can be disabled

oliverlee committed:
handle precondition error printing for multiple archs

nmm0 commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -133,11 +134,11 @@ class layout_left::mapping {
       : __extents(__other.extents())
     {
```
This should just be guarded by C++17. We shouldn't include the layout_padded_fwd.hpp header in C++14 mode

crtrott replied:
done

nmm0 commented on include/experimental/__p2642_bits/layout_padded_fwd.hpp:
```diff
@@ -82,36 +83,49 @@ struct is_layout_right_padded_mapping<_Mapping,
   std::enable_if_t<std::is_same<_Mapping, typename layout_right_padded<_Mapping::padding_value>::template mapping<typename _Mapping::extents_type>>::value>>
     : std::true_type {};
 
+
```
Let's just not include this header in C++14, since it's only used for padded layouts which require at least C++17

nmm0 commented on include/experimental/__p0009_bits/layout_right.hpp:
```diff
@@ -18,8 +18,8 @@
-#include <stdexcept>
 #include "layout_stride.hpp"
+#include "utility.hpp"
 #include "../__p2642_bits/layout_padded_fwd.hpp"
```
Same here, let's guard this with `MDSPAN_HAS_CXX_17`

nmm0 reviewed (dismissed):
I'm good with almost everything, just lets keep the padded layout stuff isolated (since it requires C++17 anyway)

crtrott committed:
Fix CUDA (and hopefully HIP/SYCL) compilation

crtrott committed:
Fix CUDA build and run issues

crtrott commented:
I just noticed that this requires CXX Extensions for GCC. I.e. -std=gnu++14 - it doesn't work with -std=c++14. Do you consider that acceptable?

crtrott committed:
More fixes for cpp14 and warnings

crtrott committed:
Guard inclusion of layout_padded_fwd

oliverlee commented:
> I just noticed that this requires CXX Extensions for GCC. I.e. -std=gnu++14 - it doesn't work with -std=c++14. Do you consider that acceptable?

Do you know what requires extensions? I didn't intend for that and I believe I am only using -std=c++14 in projects using mdspan (however, those don't use CMake).

crtrott commented:
Its now fixed.

dalg24 commented on include/experimental/__p0009_bits/layout_stride.hpp:
```diff
@@ -447,32 +451,48 @@ struct layout_stride {
+
+  public:
+    MDSPAN_INLINE_FUNCTION _MDSPAN_CONSTEXPR_14 bool is_exhaustive() const noexcept {
+      return is_exhaustive_impl(detail::with_rank<extents_type::rank()>{});
```
No warning from `is_exhaustive_impl` not having the host device markups?

crtrott replied:
we are currently using constexpr-experimental stuff for mdspan, we will fix that later.

dalg24 replied:
fine

dalg24 commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -18,7 +18,12 @@
+#include <cstdlib>
 #include <type_traits> // std::is_void
+#if defined(_MDSPAN_HAS_CUDA) || defined(_MDSPAN_HAS_HIP) || defined(_MDSPAN_HAS_SYCL)
+#include "assert.h"
```
Please confirm that you mean to include this header and not `<cassert>`

crtrott replied:
I couldn't find a combo which fails. For Kokkos we anyway gonna replace the thing with Kokkos::abort.

dalg24 commented on include/experimental/__p0009_bits/macros.hpp:
```diff
@@ -101,6 +106,69 @@
+MDSPAN_FUNCTION inline void default_precondition_violation_handler(const char* cond, const char* file, unsigned line)
+{
+  printf("%s:%u: precondition failure: `%s`\n", file, line, cond);
+  assert(0);
```
Please confirm that you did not see issue with Clang as the CUDA compiler.

crtrott replied:
Yeah I couldn't find a variant which fails and we will use in Kokkos Kokkos::abort anyway.

dalg24 commented on tests/test_layout_preconditions.cpp:
```diff
@@ -0,0 +1,32 @@
+
+#if defined(_MDSPAN_HAS_CUDA) || defined(_MDSPAN_HAS_HIP) || defined(_MDSPAN_HAS_SYCL)
+  ASSERT_DEATH(Kokkos::layout_left::mapping<E>{stride}, "");
+  ASSERT_DEATH(Kokkos::layout_right::mapping<E>{stride}, "");
```
Wait why don't we special cases these?  Is the error message lost?

crtrott replied:
Yeah because we don't use all the __assertfail crap.

dalg24 commented on tests/test_mdspan_size.cpp:
```diff
@@ -55,6 +55,7 @@ void test_mdspan_size(std::vector<char>& storage, Extents&& e)
 
 TEST(TestMdspan, MdspanSizeReturnTypeAndPrecondition)
 {
+  (void) test_info_;
```
Yuk

dalg24 commented on tests/test_macros.cpp:
```diff
@@ -0,0 +1,28 @@
+#else
+  constexpr auto msg = "hello, world!";
+#endif
+
```
I would have done
```c++
#if MDSPAN_IMPL_CHECK_PRECONDITION
ASSERT_DEATH(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"), msg);
#else
ASSERT_NO_THROW(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"));
#endif
```
instead of messing with the test success criterion in the cmake file.

dalg24 replied:
Please ignore.   This doesn't apply here.  I is only for `test_precondition_checks_can_be_disabled` Sorry about the noise.

dalg24 reviewed (approved):
Good enough.
Thank you Oliver.
